### PR TITLE
[i8042prt] Initialize mice correctly and remove the need for device hacks

### PR DESCRIPTION
**Pull request by BelleNottelling**
## Purpose

Correctly initialize mice correctly to fix issues on some laptops, then initialize the wheel again to prevent it no longer working.

JIRA issue: [CORE-6901](https://jira.reactos.org/browse/CORE-6901)
JIRA issue: [CORE-12663](https://jira.reactos.org/browse/CORE-12663)
JIRA issue: [CORE-12434](https://jira.reactos.org/browse/CORE-12434)

## Proposed changes

- During mouse detection set it into stream mode (which is what we use) and then disable error reporting (as described [here](https://isdaman.com/alsos/hardware/mouse/ps2interface.htm))
- Initialize the mouse wheel (Microsoft IntelliMouse compatible devices) by setting the polling rate to `200`, `100`, and then `80`
- If the device reports that it's a Microsoft IntelliMouse compatible device enable the scroll wheel

## TODO
- Test and verify this patch doesn't break HW (In progress)

## Tests performed
 _Checkmark means it passed_

- \[x] Latitude E6400
- \[x] Virtualbox with PS2 mouse enabled in settings
- \[x] Qemu

### CODE REVIEW

**Comment by BenNottelling** [comment]
Scrolling in qemu doesn't work so i'm assuming that's also going to effect other devices

**Comment by HBelusca** [comment]
> Please make sure, that it does not cause any regressions, i.e. it must work on all major virtualization platforms (vbox, qemu, vmware, hyper-v) and should be tested on a large range of hardware. I'd rather keep the hacks, than breaking the mouse on any system.

I think then that the hacks could be somewhat kept, but used only after the main solution proposed by @BenNottelling is run on the hardware. If the hacks need to be run then a DPRINT can be displayed to signal this, so that we can know on which hardware the hack was still needed?

**Comment by BenNottelling** [comment]
@tkreuzer I can't test hyper-v, do you know if there's anyone else who can?

**Comment by ColinFinck** [comment]
I've added this PR to https://reactos.org/wiki/ReactOS_Hackfest_2019/Lists#Ideas
It's still 1.5 months till Hackfest, but we should have a lot of hardware there to test this patch and ensure that there are no regressions. I finally want to get rid of per-device hacks in this driver and get basic keyboard/mouse support on every platform! Want to join us at Hackfest? :)

In the meantime, let's get the QEMU scrolling regression fixed and ensure proper functionality on all machines/VMs we have readily available.

**Comment by BenNottelling** [comment]
@ColinFinck I'd love to join in hackfest, but I don't really think I have the skills to do so.
(Also, I believe the trip is a bit far for me)

Here's the current status of this PR
1. When using stream mode with PS/2 mice (as we do in this driver) we need to disable error reporting. This FIXES the trackpad on my latitude E6400 WITHOUT issuing a reboot to the mouse. Yay!
2. Scrolling is breaking, from what I understand, because we need to enable the Microsoft Intellimouse Extension on devices that support it. 

So, from what I understand there, this driver should currently work on ALL PS/2 mice, but without scrolling. So that's what I need to work on next.

File: drivers/input/i8042prt/pnp.c [hunk 1]
```diff
@@ -248,6 +248,41 @@ i8042DetectMouse(
...
+	i8042IsrWritePort(DeviceExtension, MOU_DEF_SAMPLE_RATE, CTRL_WRITE_MOUSE);
+	i8042IsrWritePort(DeviceExtension, 0x50, CTRL_WRITE_MOUSE);
+	
+	i8042IsrWritePort(DeviceExtension, MOU_READ_DEV_TYPE, CTRL_WRITE_MOUSE);
+	i8042ReadDataWait(DeviceExtension, &Value);
+	//TODO: Help me! The value of MouseType doesn't actually get updated so the wheel logic never gets appplied 
+	if(Value == 0x03)
+	{
+		
+        IN I8042_MOUSE_EXTENSION DeviceExtension;
```
Comment (HBelusca) [inline drivers/input/i8042prt/pnp.c]: Wth is that "IN" thing, and why is this variable declared there and not used afterwards?

**Comment by BenNottelling** [comment]
Ok, currently
Latitude E6400 & v66: Mouse works, but jumps around the screen constantly (down and left), I'm not really sure what is happening there, but wondering if the driver is getting out of sync or something else. v86 doesn't have a scroll wheel implemented

Virtualbox: Mouse works without any issues, including scrolling

Qemu: Mouse works, but seemingly never reports the device ID as 0x03 so the mouse wheel isn't enabled in the driver, scrolling causes the mouse to jump around. 

I'll have to continue to read up on how PS/2 mice function and find a solution for those regressions.

**Comment by BenNottelling** [comment]
This patch is complete and I consider it ready to be reviewed and tested for regressions

(Check is a pass)
Qemu:
- \[x] Movement
- \[x] Scrolling

Virtualbox:
- \[x] Movement
- \[x] Scrolling

VMware player:
- \[x] Movement
- \[x] Scrolling

Latitude E6400:
- \[x] Movement
- \[x] Scrolling (No scroll wheel, I think) 

Here's a link with both a bootCD and a liveCD using this driver so everyone can easily test it. 
https://drive.google.com/drive/folders/1KsipImctEWV94WEtXXkkXA8qf3HDFgbX?usp=sharing

**Comment by encodedpr** [comment]
What reference material are you using to arrive at these modifications?

**Comment by BenNottelling** [comment]
@encodedpr [This](https://isdaman.com/alsos/hardware/mouse/ps2interface.htm) is my reference, I'll also breakdown how/why this works

First, after detecting the mouse we send the command to change it to streaming mode, which is what this driver is written for
If the mouse isn't in streaming mode, it'll wait for a data packet to be requested, but that would never come. (it seems that usually these PS/2 mice default to steaming mode though)
Next, we disable error reporting as any error messages in our stream would cause confusion and de-sync  on the driver side.

This fixes general mice usage, but does not allow the scroll wheel to function, which is the next part we do, this is also the part that looks ridiculous if you don't have any understanding of what's happening

Standard PS/2 mice used to not have scroll wheels, and that is instead something that Microsoft created, the Intellimouse extension, which is initialized by the following:

Setting polling rate to 200
Setting polling rate to 100
Setting polling rate to 80
Request device ID

In the event that we have a standard PS/2 mouse, this will do no harm, and return the device ID as `0x00` with us setting the polling rate back to the standard 100 after the check
If we have a mouse with the Intellimouse extension it will return a device ID of `0x03` and enable the scroll wheel. 

This code for checking the device ID is in a while statement for an important reason
```
	while(chkcnt < 10) {
	    i8042ReadDataWait(DeviceExtension, &Value);

	    if(Value == 0x03)
	    {
	        // This is an Intellimouse or Intellimouse Explorer
	        DeviceExtension->MouseExtension->MouseType = Intellimouse;
	        chkcnt=11;
	    }
	    else
	    {
	        // Regular PS/2 Mouse
	        DeviceExtension->MouseExtension->MouseType = GenericPS2;
	    }
	    chkcnt++;
	}
```
Initially I just checked for the device ID right after requesting the device ID, which worked under Virtualbox but NOT qemu, the reason for this was because qemu outputted responses when we set the polling rate, and Virtualbox did not, so reading the data from the mouse on Vbox got me a 0x03 (device ID) and qemu gave me a acknowledgement response instead. 

To fix this, we read from the mouse 10 times, or until we read `0x03`
On Virtualbox we immediately get `0x03` so we set the mouse type to `Intellimouse` which enables the scroll wheel and set `chkcnt` to 11 which ends the while loop

In qemu it has to read from the stream a couple times before we get to the response with our device ID, at which point we set `chkcnt` to 11, ending the check and continue in the code

I used 10 as the number of times to check as that felt like a good middle ground, giving enough checks to get through responses from setting the polling rate, and not enough to really slow down the device initialization in the event that we have a standard mouse :)

Hope that answered any questions you had!

File: drivers/input/i8042prt/pnp.c [hunk 2]
```diff
@@ -248,6 +249,43 @@ i8042DetectMouse(
...
+	i8042IsrWritePort(DeviceExtension, MOU_READ_DEV_TYPE, CTRL_WRITE_MOUSE);
+		
+	while(chkcnt < 10) {
+	    i8042ReadDataWait(DeviceExtension, &Value);
+		
+	    if(Value == 0x03)
+	    {
+	        // This is an Intellimouse or Intellimouse Explorer
+	        DeviceExtension->MouseExtension->MouseType = Intellimouse;
+	        chkcnt=11;
```
Comment (binarymaster) [inline drivers/input/i8042prt/pnp.c]:
```suggestion
	        break;
```

**Comment by BenNottelling** [comment]
Currently the first stage of the BootCD is broken because `DeviceExtension->MouseExtension` isn't being populated, works fine on a LiveCD though

File: drivers/input/i8042prt/pnp.c [hunk 3]
```diff
@@ -263,6 +264,44 @@ i8042DetectMouse(
...
+
+    i8042IsrWritePort(DeviceExtension, MOU_DEF_SAMPLE_RATE, CTRL_WRITE_MOUSE);
+    i8042IsrWritePort(DeviceExtension, 0x64, CTRL_WRITE_MOUSE);
+	
+    i8042IsrWritePort(DeviceExtension, MOU_DEF_SAMPLE_RATE, CTRL_WRITE_MOUSE);
+    i8042IsrWritePort(DeviceExtension, 0x50, CTRL_WRITE_MOUSE);
+
+    i8042IsrWritePort(DeviceExtension, MOU_READ_DEV_TYPE, CTRL_WRITE_MOUSE);
+
+    while(chkcnt < 10) {
```
Comment (HBelusca) [inline drivers/input/i8042prt/pnp.c]: Fix code style please.

File: drivers/input/i8042prt/pnp.c [hunk 4]
```diff
@@ -263,6 +264,44 @@ i8042DetectMouse(
...
+        i8042ReadDataWait(DeviceExtension, &Value);
+		
+        if(Value == 0x03)
```
Comment (HBelusca) [inline drivers/input/i8042prt/pnp.c]: ditto.

File: drivers/input/i8042prt/pnp.c (hunk 3 above)
Comment (HBelusca) [reply drivers/input/i8042prt/pnp.c]: Also you can use a for-loop instead.

Comment (BenNottelling) [reply drivers/input/i8042prt/pnp.c]: Is a for - loop more efficient? 
(or anything else)

Comment (BenNottelling) [reply drivers/input/i8042prt/pnp.c]: Actually, I just tried to use a for-loop, and it broke the mouse logic

**Comment by BenNottelling** [comment]
> Note: the MouseType is also set in https://git.reactos.org/?p=reactos.git;a=blob;f=drivers/input/i8042prt/mouse.c;hb=6d238e451345dfd4f5b8178d4d515801546fe95d#l659 , so please check why you also had to do it in the new place.

I'll have to read further into that function to find out why it's not getting set in like it should. 

I rebased with current master and force pushed everything so we're back to a single clean commit. 
Overall, everyone has been reporting good things when testing the driver, MOSTLY. 
@BieHDC Reported no regressions, but also no fixes for himself
A couple people commented on the JIRA issues saying it worked for them
Mattermost user Zombiedeth reported this fixing his scroll wheel on his desktop 

But. Zombiedeth also said that his `Compaq CQ2000` no longer boots, and I can confirm that v86 also seems to be broken with this driver
I'm not sure why this would present issues booting, is there a memory leak somewhere? Not sure where / how I would diagnose this issue.

File: drivers/input/i8042prt/pnp.c (hunk 3 above)
Comment (HBelusca) [reply drivers/input/i8042prt/pnp.c]: > Is a for-loop more efficient?
> (or anything else)

No, but it would be more readable.

Comment (binarymaster) [reply drivers/input/i8042prt/pnp.c]: How about that... :slightly_smiling_face: 

```
int chkcnt = 10;

while (chkcnt--)
{
    ...
}
```

File: drivers/input/i8042prt/pnp.c [hunk 5]
```diff
@@ -263,6 +263,47 @@ i8042DetectMouse(
...
+    i8042IsrWritePort(DeviceExtension, 0xC8, CTRL_WRITE_MOUSE);
+
+    i8042IsrWritePort(DeviceExtension, MOU_DEF_SAMPLE_RATE, CTRL_WRITE_MOUSE);
+    i8042IsrWritePort(DeviceExtension, 0x64, CTRL_WRITE_MOUSE);
+	
+    i8042IsrWritePort(DeviceExtension, MOU_DEF_SAMPLE_RATE, CTRL_WRITE_MOUSE);
+    i8042IsrWritePort(DeviceExtension, 0x50, CTRL_WRITE_MOUSE);
+
+    i8042IsrWritePort(DeviceExtension, MOU_READ_DEV_TYPE, CTRL_WRITE_MOUSE);
+    while(chkcnt--)
```
Comment (Extravert-ir) [inline drivers/input/i8042prt/pnp.c:287]: Why not a for loop?

File: drivers/input/i8042prt/pnp.c [hunk 6]
```diff
@@ -263,6 +263,47 @@ i8042DetectMouse(
...
+    while(chkcnt--) 
+	{
+        i8042ReadDataWait(DeviceExtension, &Value);
+        if(Value == 0x03)
+        {
+            DeviceExtension->MouseExtension->MouseType = Intellimouse;
+            break;
+        }
+        else if(Value == 0x00)
+		{
```
Comment (Extravert-ir) [inline drivers/input/i8042prt/pnp.c:296]: tabs instead of spaces

**Pull request closed without merging**